### PR TITLE
Enable separate EKF2 instances for each IMU

### PR DESCRIPTION
**Pull request by priseborough**
EKF is split into front and backend so it does not use memory when deactivated using EKF_ENABLE
An EK2_IMU_MASK variable is added which specifies which IMU's will be used by EKF2 and therefore the number of instances of EKF2. A separate instance of EKF2 is started for each IMU selected.
The switching between IMU data for EKF2 has been removed and replaced by selection of EKF2 outputs, where each EKF2 is using a different IMU.
Selection of EKF2 outputs is made using a sticky health check and the health check has been made more sensitive with an additional criteria to detect loss of attitude reference. 
EKF messages are only logged when EKF_ENABLE is non zero.
The innovations are logged for additional EKF2 instances

The code has been tested in SITL and does switch when significant IMU errors are introduced, however the switch is not instant and could be improved with addition of a first stage IMU comparison check followed by a second stage innovation consistency check.

This code has been flown on a copter wth the following param settings:

TERRAIN_ENABLE = 0
EKF_ENABLE = 0
AHRS_EKF_TYPE = 2
EK2_ENABLE = 1
EK2_IMU_MASK = 3

### CODE REVIEW

**Comment by tridge** [comment]
Paul and I have discussed this and we've agreed on a different approach to handling the healthy() function. A quality metric will be built within NavEKF2 for selecting which core to use, and external healthy status won't go to false if the EK2 code can still provide a solution

**Comment by priseborough** [comment]
The PR has been re-worked and i have demonstrated a clean switchover in SITL after injecting a 1 rad/sec error into the roll gyro or a 10 m/s/s error into the Y or Z accelerometer.

File: libraries/AP_NavEKF2/AP_NavEKF2.cpp [hunk 1]
```diff
@@ -515,6 +525,18 @@ bool NavEKF2::healthy(void) const
     return core[primary].healthy();
 }
 
+// returns the index of the primary core
+// return false if no primary core selected
+bool NavEKF2::getPrimaryCoreIndex(uint8_t index) const
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp]: I think you want index to be a reference?

Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp]: it may be better to make this return a int8_t, and use -1 for no core

Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2.cpp]: Yes, that makes sense. I'll update the PR.

File: libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp [hunk 2]
```diff
@@ -123,7 +123,7 @@ void NavEKF2_core::setAidingMode()
             stateStruct.position.z = -meaHgtAtTakeOff;
         } else if (frontend->_fusionModeGPS == 3) {
             // We have commenced aiding, but GPS useage has been prohibited so use optical flow only
-            hal.console->printf("EKF2 is using optical flow\n");
+            hal.console->printf("EKF2 IMU%i is using optical flow\n",imu_index);
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp]: this should be IMU%u and cast imu_index to unsigned, like this:
  hal.console->printf("EKF2 IMU%u is using optical flow\n", (unsigned)imu_index);
what you've done will work, but isn't as portable

**Comment by priseborough** [comment]
I have submitted patches to address the review comments and  have also completed the to-do items.

**Comment by priseborough** [comment]
This PR now includes full logging of the second instance of EKF2 (if enabled). Due to size limitations, optical flow debug data is only logged for the primary instance.

**Comment by tridge** [comment]
applied, thanks Paul!

**Pull request closed without merging**